File: src/main/java/com/developersam/web/model/scheduler/Scheduler.java
package com.developersam.web.model.scheduler;

import com.developersam.web.util.DataStoreObject;
import com.developersam.web.util.DateUtil;
import com.google.appengine.api.datastore.PreparedQuery;
import com.google.appengine.api.datastore.Query;
import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
import com.google.appengine.api.datastore.Query.Filter;
import com.google.appengine.api.datastore.Query.FilterOperator;
import com.google.appengine.api.datastore.Query.FilterPredicate;
import com.google.appengine.api.datastore.Query.SortDirection;
import com.google.appengine.api.users.UserServiceFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import static com.google.appengine.api.datastore.Query.FilterOperator.EQUAL;
import static com.google.appengine.api.datastore.Query.FilterOperator.GREATER_THAN_OR_EQUAL;

/**
 * Represent the scheduler app.
 */
public class Scheduler extends DataStoreObject {
    
    /**
     * The only instance of itself.
     */
    public static final Scheduler INSTANCE = new Scheduler();
    
    /**
     * A private default no-arg constructor.
     */
    private Scheduler() {
        super("SchedulerItem");
    }
    
    /**
     * Obtain a list of all scheduler items for a user signed in.
     *
     * @return a list of scheduler items
     */
    public List<SchedulerItem> getAllSchedulerItems() {
        String userEmail = UserServiceFactory.getUserService()
                .getCurrentUser().getEmail();
        Filter filterUser = new FilterPredicate(
                "userEmail", EQUAL, userEmail);
        Filter filterDeadline = new FilterPredicate(
                "deadline", GREATER_THAN_OR_EQUAL,
                DateUtil.getYesterday());
        List<Boolean> trueAndFalse = new ArrayList<>(2);
        trueAndFalse.add(true);
        trueAndFalse.add(false);
        // Just to overcome Datastore's indexing and sorting limitation.
        Filter filterCompleted = new FilterPredicate(
                "completed", FilterOperator.IN, trueAndFalse);
        Filter filter = CompositeFilterOperator.and(
                filterCompleted, filterUser, filterDeadline);
        Query q = getQuery().
                setFilter(filter).
                addSort("completed", SortDirection.ASCENDING).
                addSort("deadline", SortDirection.ASCENDING);
        PreparedQuery pq = getPreparedQuery(q);
        return StreamSupport.stream(pq.asIterable().spliterator(), false)
                .map(SchedulerItem::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Delete a scheduler item with a given key.
     *
     * @param key key of the item to be deleted.
     */
    public void delete(String key) {
        SchedulerItem item = SchedulerItem.from(key);
        if (item != null) {
            item.delete();
        }
    }
    
    /**
     * Change the completion status for a given scheduler item.
     *
     * @param key key of the item.
     * @param complete completion status.
     */
    public void changeCompletionStatus(String key, boolean complete) {
        SchedulerItem item = SchedulerItem.from(key);
        if (item != null) {
            item.markAs(complete);
        }
    }
    
}

File: src/main/kotlin/com/developersam/model/scheduler/Scheduler.kt
package com.developersam.model.scheduler

import com.developersam.util.datastore.DataStoreObject
import com.developersam.util.datastore.dataStore
import com.developersam.util.yesterday
import com.google.appengine.api.datastore.Query.CompositeFilterOperator
import com.google.appengine.api.datastore.Query.FilterOperator
import com.google.appengine.api.datastore.Query.FilterPredicate
import com.google.appengine.api.datastore.Query.SortDirection
import com.google.appengine.api.users.UserServiceFactory
import java.util.ArrayList
import java.util.stream.StreamSupport

/**
 * Represent the scheduler app.
 */
object Scheduler : DataStoreObject(kind = "SchedulerItem") {

    /**
     * Obtain a list of [allSchedulerItems] for a signed-in user.
     */
    val allSchedulerItems: Array<SchedulerItem>
        get() {
            val userEmail = UserServiceFactory.getUserService()
                    .currentUser.email
            val filterUser = FilterPredicate("userEmail",
                    FilterOperator.EQUAL, userEmail)
            val filterDeadline = FilterPredicate("deadline",
                    FilterOperator.GREATER_THAN_OR_EQUAL, yesterday)
            val trueAndFalse = ArrayList<Boolean>(2)
            trueAndFalse.add(true)
            trueAndFalse.add(false)
            val filterCompleted = FilterPredicate(
                    "completed", FilterOperator.IN, trueAndFalse)
            val filter = CompositeFilterOperator.and(
                    filterCompleted, filterUser, filterDeadline)
            val q = query.setFilter(filter)
                    .addSort("completed", SortDirection.ASCENDING)
                    .addSort("deadline", SortDirection.ASCENDING)
            // :< Just to overcome Datastore's indexing and sorting limitation.
            val pq = dataStore.prepare(q)
            return StreamSupport.stream(
                    pq.asIterable().spliterator(), false)
                    .map({ SchedulerItem(it) })
                    .toArray({ size -> arrayOfNulls<SchedulerItem>(size) })
        }

    /**
     * Delete a scheduler item with a given [key].
     */
    fun delete(key: String) {
        val item = SchedulerItem.from(key)
        item?.deleteFromDatabase()
    }

    /**
     * Mark the completion status for a scheduler item specified by the
     * given [key] and the desired new [completionStatus].
     */
    fun markAs(key: String, completionStatus: Boolean) {
        val item = SchedulerItem.from(key)
        item?.markAs(completionStatus)
    }

}